### PR TITLE
removes the clunky calls to GetService and hides IConfigureOptions<T> /8 data-protect-apis/ra

### LINKED ISSUE #17643: Data Protection - Provide IXmlRepository with singleton dependencies
In a .NET Core 3.1 project, one may configure Data Protection in `ConfigureServices` like below:

```
public void ConfigureServices(IServiceCollection services)
{
    // ...

    var sp = services.BuildServiceProvider();
    services.AddDataProtection()
            .AddKeyManagementOptions(o => o.XmlRepository = sp.GetService<IXmlRepository>());
}
```

This raises below warning (which makes sense):

> Calling 'BuildServiceProvider' from application code results in an additional copy of singleton services being created. Consider alternatives such as dependency injecting services as parameters to 'Configure'.

I'm using a service provider there because my implementation of `IXmlRepository` has a dependency to a singleton service.

What would be the recommended way of providing my `IXmlRepository` implementation without having to build the service provider (and thus making additional copies of singleton services)?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 532d66a6-fa01-54d3-1dfa-20d1889668bb
* Version Independent ID: 6489c522-2148-e39a-92cc-6f631c065422
* Content: [Get started with the Data Protection APIs in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/using-data-protection?view=aspnetcore-3.1)
* Content Source: [aspnetcore/security/data-protection/using-data-protection.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/security/data-protection/using-data-protection.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #17643
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/security/data-protection/using-data-protection?view=aspnetcore-5.0&branch=pr-en-us-22287)

To review the public build 
- download https://www.dropbox.com/sh/yy994551cje0kap/AAB0tRnm7gze2kYrSYPSFHIda?dl=0
- unzip
- view HTML file 
- 
@AlissonRS @csalgo70 @ryan-morris @stefanb05 @maulik-modi please review

This is the Razor Pages template with the following additions:
image

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@AlissonRS @csalgo70 @ryan-morris @stefanb05 @maulik-modi please review

**Comment by ryan-morris** [comment]
This description made sense to me. If that had been in the docs I wouldn't have ended up up on the original issue. I started with a similar approach before realizing `Microsoft.AspNetCore.DataProtection.EntityFrameworkCore` was a thing.

**Comment by maulik-modi** [comment]
@ryan-morris , Do you mean EFCore is the mandate?

File: aspnetcore/security/data-protection/using-data-protection.md [hunk 1]
```diff
@@ -29,3 +32,32 @@ When you create a protector you must provide one or more [Purpose Strings](xref:
...
+
+The preceding code logs the following warning:
+
+  *Calling 'BuildServiceProvider' from application code results in an additional copy of singleton services being created. Consider alternatives such as dependency injecting services as parameters to 'Configure'.*
+
+The following code provides the `IXmlRepository` implementation without having to build the service provider and therefore making additional copies of singleton services:
+
+```csharp
+services.AddOptions<KeyManagementOptions>()
+        .Configure<IMyDataStore>((options, store) =>
```
Comment (AlissonRS) [inline aspnetcore/security/data-protection/using-data-protection.md:57]: I'd suggest making it more explicit the dependency passed to this `Configure<>()` is not supposed to be scoped (hence you shouldn't inject a `DbContext`), instead you should inject some factory like `IServiceScopeFactory` and use it in your `IXmlRepository` implementation to create a new scope and get scoped dependencies like `DbContext` if you need.

**Comment by Rick-Anderson** [comment]
@AlissonRS can you create the simplest project that shows how you code works and post it on GitHub? We can use that sample.

File: aspnetcore/security/data-protection/using-data-protection/samples/CustomXMLrepo/CustomXMLrepo/Pages/Index.cshtml.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            _logger = logger;
+            _context = dataProtectionDbContext;
+        }
+
+        public IList<XmlKey> XmlKey { get; set; }
+
+        public async Task<Microsoft.AspNetCore.Mvc.ContentResult> OnGetAsync()
+        {
+            XmlKey = await _context.XmlKeys.ToListAsync();
+            return Content("Key Count = " + XmlKey.Count().ToString());
```
Comment (Rick-Anderson) [inline aspnetcore/security/data-protection/using-data-protection/samples/CustomXMLrepo/CustomXMLrepo/Pages/Index.cshtml.cs:28]: @AlissonRS I need to clean this up but it does show 1 key.

**Comment by Rick-Anderson** [comment]
@AlissonRS 
To review the public build 
- download https://www.dropbox.com/sh/yy994551cje0kap/AAB0tRnm7gze2kYrSYPSFHIda?dl=0
- unzip
- view HTML file

**Comment by AlissonRS** [comment]
@Rick-Anderson it's looking good, thanks for that. I know I'm a bit late, but would you still like me to create the simplest project showing how to use it with EF Core? I can do it today if you need, please let me know.

**Review by AlissonRS — APPROVED** [review]

**Comment by HaoK** [comment]
Looks great, I wonder if we should reference this from the options documents in some way as well, since I don't think the OptionsBuilder has enough documentation about this style of API that we added fairly late.  Its subtle and sort of impossible to discover today by developers on their own

**Review by HaoK — APPROVED** [review]

File: aspnetcore/security/data-protection/using-data-protection.md [hunk 3]
```diff
@@ -29,3 +32,38 @@ When you create a protector you must provide one or more [Purpose Strings](xref:
 > Instances of `IDataProtectionProvider` and `IDataProtector` are thread-safe for multiple callers. It's intended that once a component gets a reference to an `IDataProtector` via a call to `CreateProtector`, it will use that reference for multiple calls to `Protect` and `Unprotect`.
 >
 >A call to `Unprotect` will throw CryptographicException if the protected payload cannot be verified or deciphered. Some components may wish to ignore errors during unprotect operations; a component which reads authentication cookies might handle this error and treat the request as if it had no cookie at all rather than fail the request outright. Components which want this behavior should specifically catch CryptographicException instead of swallowing all exceptions.
+
+<a name="add-opt"></a>
+
+## Use AddOptions to configure custom repository
```
Comment (Rick-Anderson) [inline aspnetcore/security/data-protection/using-data-protection.md:38]: @AlissonRS please suggest a better title.

File: aspnetcore/fundamentals/configuration/options.md [hunk 4]
```diff
@@ -132,6 +132,8 @@ All options are named instances. <xref:Microsoft.Extensions.Options.IConfigureOp
 
 `OptionsBuilder` is used in the [Options validation](#val) section.
 
+See [Use AddOptions to configure custom repository](xref:security/data-protection/using-data-protection#add-opt) for information adding a custom repository.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/configuration/options.md:135]: @AlissonRS please suggest better wording.

**Pull request merged**